File: src/main/java/me/taubsie/dungeonhub/application/command/commands/CarryTypeCommand.java
package me.taubsie.dungeonhub.application.command.commands;

import me.taubsie.dungeonhub.application.command.Command;
import me.taubsie.dungeonhub.application.command.CommandParameters;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.CarryTypeConnection;
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionException;
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException;
import me.taubsie.dungeonhub.application.exceptions.InvalidSubCommandException;
import me.taubsie.dungeonhub.application.service.ApplicationService;
import me.taubsie.dungeonhub.common.model.carry_type.CarryTypeCreationModel;
import me.taubsie.dungeonhub.common.model.carry_type.CarryTypeModel;
import me.taubsie.dungeonhub.common.model.carry_type.CarryTypeUpdateModel;
import org.javacord.api.entity.DiscordEntity;
import org.javacord.api.entity.channel.ChannelType;
import org.javacord.api.entity.channel.ServerChannel;
import org.javacord.api.entity.permission.PermissionType;
import org.javacord.api.entity.server.Server;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.SlashCommandInteractionOption;
import org.javacord.api.interaction.SlashCommandOption;
import org.javacord.api.interaction.SlashCommandOptionBuilder;
import org.javacord.api.interaction.SlashCommandOptionType;

import java.util.List;
import java.util.Optional;

@CommandParameters(name = "carry-type", description = "Set up the carry types for this server.",
        enabledForPermissions = PermissionType.ADMINISTRATOR)
public class CarryTypeCommand extends Command {
    public static final String FIELD_NAME = "carry-type";

    public static SlashCommandOption getCarryTypeOption() {
        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName(FIELD_NAME)
                .setDescription("The identifier of the carry type")
                .setRequired(true)
                .setMaxLength(30)
                .setAutocompletable(true)
                .build();
    }

    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        SlashCommandInteractionOption subCommand = getOptionAtIndex(0);

        switch (subCommand.getName().toLowerCase()) {
            case "create" -> create(subCommand);
            case "delete" -> delete(subCommand);
            case "edit" -> edit(subCommand);
            case "get" -> get(subCommand);
            case "reset" -> reset(subCommand);
            default -> throw new InvalidSubCommandException();
        }
    }

    public void create(SlashCommandInteractionOption subCommand) {
        Server server = getServer();
        String identifier = getStringOption(subCommand, "identifier")
                .strip()
                .toLowerCase()
                .replace(" ", "_");
        String displayName = getStringOption(subCommand, "display-name");

        if (CarryTypeConnection.getInstance(server.getId()).getByIdentifier(identifier).isPresent()) {
            throw new InvalidOptionException("identifier", "That carry type already exists!");
        }

        CarryTypeCreationModel creationModel = new CarryTypeCreationModel(identifier, displayName);

        getOptionalChannelOption(subCommand, "log-channel")
                .ifPresent(channel -> creationModel.setLogChannel(channel.getId()));
        getOptionalChannelOption(subCommand, "leaderboard-channel")
                .ifPresent(channel -> creationModel.setLeaderboardChannel(channel.getId()));
        getOptionalBooleanOption(subCommand, "event-active")
                .ifPresent(creationModel::setEventActive);

        Optional<CarryTypeModel> carryTypeModel = CarryTypeConnection.getInstance(server.getId())
                .addNewCarryType(creationModel);

        if (carryTypeModel.isEmpty()) {
            //TODO custom class?
            throw new CommandExecutionException("Couldn't add that carry type.");
        }

        respond(ApplicationService.getInstance()
                .getCarryTypeEmbed(carryTypeModel.get())
                .setTitle("Carry Type created"));
    }

    public void delete(SlashCommandInteractionOption subCommand) {
        String identifier = getStringOption(subCommand, FIELD_NAME);

        Optional<CarryTypeModel> carryType = CarryTypeConnection.getInstance(getServer().getId())
                .getByIdentifier(identifier);

        if (carryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("That carry type doesn't exists!");
        }

        Optional<CarryTypeModel> deletedCarryType = CarryTypeConnection.getInstance(carryType.get().getServer().getId())
                .deleteCarryType(carryType.get());

        if (deletedCarryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("Carry type couldn't be deleted!");
        }

        respond(ApplicationService.getInstance()
                .getCarryTypeEmbed(deletedCarryType.get())
                .setTitle("Deleted Carry Type"));
    }

    public void edit(SlashCommandInteractionOption editCommand) {
        Optional<CarryTypeModel> carryType = CarryTypeConnection.getInstance(getServer().getId())
                .getByIdentifier(getStringOption(editCommand, FIELD_NAME));

        if (carryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("That carry type doesn't exists!");
        }

        Optional<String> displayName = getOptionalStringOption(editCommand, "display-name");
        Optional<ServerChannel> logChannel = getOptionalChannelOption(editCommand, "log-channel");
        Optional<ServerChannel> leaderboardChannel = getOptionalChannelOption(editCommand, "leaderboard-channel");
        Optional<Boolean> eventActive = getOptionalBooleanOption(editCommand, "event-active");

        if (displayName.isEmpty() && logChannel.isEmpty() && leaderboardChannel.isEmpty() && eventActive.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("Please provide something you want to edit.");
        }

        CarryTypeUpdateModel updateModel = new CarryTypeUpdateModel();

        displayName.ifPresent(updateModel::setDisplayName);
        logChannel.map(DiscordEntity::getId).ifPresent(updateModel::setLogChannel);
        leaderboardChannel.map(DiscordEntity::getId).ifPresent(updateModel::setLeaderboardChannel);
        eventActive.ifPresent(updateModel::setEventActive);

        Optional<CarryTypeModel> updatedCarryType = CarryTypeConnection.getInstance(getServer().getId())
                .updateCarryType(carryType.get().getId(), updateModel);

        if (updatedCarryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("Couldn't update carry type.");
        }

        respond(ApplicationService.getInstance()
                .getCarryTypeEmbed(updatedCarryType.get())
                .setTitle("Updated Carry Type"));
    }

    public void reset(SlashCommandInteractionOption resetCommand) {
        Optional<CarryTypeModel> carryType = CarryTypeConnection.getInstance(getServer().getId())
                .getByIdentifier(getStringOption(resetCommand, FIELD_NAME));

        if (carryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("That carry type doesn't exists!");
        }

        Boolean logChannel = getBooleanOption(resetCommand, "log-channel");
        Boolean leaderboardChannel = getBooleanOption(resetCommand, "leaderboard-channel");

        if (Boolean.FALSE.equals(logChannel) && Boolean.FALSE.equals(leaderboardChannel)) {
            //TODO custom class
            throw new CommandExecutionException("Please provide something you want to reset.");
        }

        CarryTypeUpdateModel updateModel = new CarryTypeUpdateModel();

        if (Boolean.TRUE.equals(logChannel)) {
            updateModel.setLogChannel(-1L);
        }

        if (Boolean.TRUE.equals(leaderboardChannel)) {
            updateModel.setLeaderboardChannel(-1L);
        }

        Optional<CarryTypeModel> updatedCarryType = CarryTypeConnection.getInstance(getServer().getId())
                .updateCarryType(carryType.get().getId(), updateModel);

        if (updatedCarryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException("Couldn't update carry type.");
        }

        respond(ApplicationService.getInstance()
                .getCarryTypeEmbed(updatedCarryType.get())
                .setTitle("Updated Carry Type with reset values"));
    }

    public void get(SlashCommandInteractionOption subCommand) {
        Optional<CarryTypeModel> carryType = CarryTypeConnection.getInstance(getServer().getId())
                .getByIdentifier(getStringOption(subCommand, FIELD_NAME));

        if (carryType.isEmpty()) {
            //TODO custom exception class
            throw new CommandExecutionException("Carry type not found.");
        }

        respondEphemeral(ApplicationService.getInstance().getCarryTypeEmbed(carryType.get()));
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        return List.of(getCreateCommand(), getDeleteCommand(), getGetCommand(), getEditCommand(), getResetCommand());
    }

    private SlashCommandOption getCreateCommand() {
        SlashCommandOption identifierOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("identifier")
                .setDescription("The identifier of the carry type.")
                .setRequired(true)
                .setMaxLength(30)
                .build();

        SlashCommandOption displayNameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("display-name")
                .setDescription("The display name of the carry type")
                .setRequired(true)
                .setMaxLength(30)
                .build();

        SlashCommandOption logChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("log-channel")
                .setDescription("Set the channel that will be used for logging")
                .setRequired(false)
                .build();

        SlashCommandOption leaderboardChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("leaderboard-channel")
                .setDescription("Set the channel that will be used to show a static leaderboard")
                .setRequired(false)
                .build();

        SlashCommandOption eventActiveOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("event-active")
                .setDescription("Set if there if an active event for score")
                .setRequired(false)
                .build();

        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("create")
                .setDescription("Create a new carry type")
                .setOptions(List.of(identifierOption, displayNameOption, logChannelOption, leaderboardChannelOption,
                        eventActiveOption))
                .build();
    }

    private SlashCommandOption getGetCommand() {
        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("get")
                .setDescription("Get information about a carry type")
                .setOptions(List.of(getCarryTypeOption()))
                .build();
    }

    private SlashCommandOption getDeleteCommand() {
        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("delete")
                .setDescription("Delete a carry type")
                .setOptions(List.of(getCarryTypeOption()))
                .build();
    }

    private SlashCommandOption getEditCommand() {
        SlashCommandOption logChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("log-channel")
                .setDescription("Set the channel that will be used for logging")
                .setRequired(false)
                .build();

        SlashCommandOption leaderboardChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("leaderboard-channel")
                .setDescription("Set the channel that will be used to show a static leaderboard")
                .setRequired(false)
                .build();

        SlashCommandOption displayNameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("display-name")
                .setDescription("Set the display name")
                .setRequired(false)
                .build();

        SlashCommandOption eventActiveOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("event-active")
                .setDescription("Set if there if an active event for score")
                .setRequired(false)
                .build();

        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("edit")
                .setDescription("Edit a carry type")
                .setOptions(List.of(getCarryTypeOption(), displayNameOption, logChannelOption,
                        leaderboardChannelOption, eventActiveOption))
                .build();
    }

    private SlashCommandOption getResetCommand() {
        SlashCommandOption logChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("log-channel")
                .setDescription("Set if the log channel should be reset")
                .setRequired(true)
                .build();

        SlashCommandOption leaderboardChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("leaderboard-channel")
                .setDescription("Set if the leaderboard channel should be reset")
                .setRequired(true)
                .build();

        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("reset")
                .setDescription("Reset properties of a carry type")
                .setOptions(List.of(getCarryTypeOption(), logChannelOption, leaderboardChannelOption))
                .build();
    }
}

File: src/main/kotlin/me/taubsie/dungeonhub/application/commands/CarryTypeCommand.kt
package me.taubsie.dungeonhub.application.commands

import dev.kord.common.entity.ChannelType
import dev.kord.common.entity.Permission
import dev.kord.common.entity.Permissions
import dev.kordex.core.commands.Arguments
import dev.kordex.core.commands.application.slash.publicSubCommand
import dev.kordex.core.commands.converters.impl.*
import dev.kordex.core.extensions.Extension
import dev.kordex.core.extensions.publicSlashCommand
import dev.kordex.core.i18n.toKey
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionWarning
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException
import me.taubsie.dungeonhub.application.loader.LoadExtension
import me.taubsie.dungeonhub.application.service.ApplicationService
import me.taubsie.dungeonhub.application.service.AutoCompletionService
import net.dungeonhub.connection.CarryTypeConnection
import net.dungeonhub.model.carry_type.CarryTypeCreationModel
import java.util.*

/**
 * Command to manage carry types.
 * This command allows the user to create, delete, get and edit carry types.
 * The user can also reset the log channel and leaderboard channel of a carry type.
 * The user can also set if an event is active for a carry type.
 * It has the following subcommands:
 * - `create`: Create a new carry type
 * - `delete`: Delete a carry type
 * - `get`: Get information about a carry type
 * - `edit`: Edit a carry type
 * - `reset`: Reset properties of a carry type
 */
@LoadExtension
class CarryTypeCommand : Extension() {
    override val name = "carry-type-command"

    override suspend fun setup() {
        publicSlashCommand {
            name = "carry-type".toKey()
            description = "Set up the carry types for this server.".toKey()
            defaultMemberPermissions = Permissions(Permission.Administrator)
            allowInDms = false

            publicSubCommand(::CarryTypeCreateArguments) {
                name = "create".toKey()
                description = "Create a new carry type".toKey()

                action {
                    respond {
                        val identifier = arguments.identifier
                            .trim()
                            .lowercase(Locale.getDefault())
                            .replace(" ", "_")

                        if (CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(identifier) != null) {
                            throw InvalidOptionException("identifier", "That carry type already exists!")
                        }

                        val creationModel = CarryTypeCreationModel(identifier, arguments.displayName)

                        if (arguments.logChannel != null) {
                            creationModel.logChannel = arguments.logChannel!!.id.value.toLong()
                        }

                        if (arguments.leaderboardChannel != null) {
                            creationModel.leaderboardChannel = arguments.leaderboardChannel!!.id.value.toLong()
                        }

                        if (arguments.eventActive != null) {
                            creationModel.eventActive = arguments.eventActive
                        }

                        val carryTypeModel =
                            CarryTypeConnection[guild!!.id.value.toLong()].addNewCarryType(creationModel)
                                ?: throw CommandExecutionWarning("Couldn't add that carry type.")

                        val embed = ApplicationService.getCarryTypeEmbed(carryTypeModel)
                        embed.title = "Carry Type created"
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTypeArguments) {
                name = "delete".toKey()
                description = "Delete a carry type".toKey()

                action {
                    respond {
                        val identifier = arguments.carryType

                        val carryType = CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(identifier)
                            ?: throw CommandExecutionWarning("That carry type doesn't exists!")

                        val deletedCarryType = CarryTypeConnection[carryType.server.id].deleteCarryType(carryType)
                            ?: throw CommandExecutionWarning("Carry type couldn't be deleted!")

                        val embed = ApplicationService.getCarryTypeEmbed(deletedCarryType)
                        embed.title = "Deleted Carry Type"
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTypeArguments) {
                name = "get".toKey()
                description = "Get information about a carry type".toKey()

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("Carry type not found.")

                        val embed = ApplicationService.getCarryTypeEmbed(carryType)
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTypeEditArguments) {
                name = "edit".toKey()
                description = "Edit a carry type".toKey()

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("That carry type doesn't exists!")

                        if (arguments.displayName == null && arguments.logChannel == null && arguments.leaderboardChannel == null && arguments.eventActive == null) {
                            throw CommandExecutionWarning("Please provide something you want to edit.")
                        }

                        val updateModel = carryType.getUpdateModel()

                        if (arguments.displayName != null) {
                            updateModel.displayName = arguments.displayName
                        }

                        if (arguments.logChannel != null) {
                            updateModel.logChannel = arguments.logChannel!!.id.value.toLong()
                        }

                        if (arguments.leaderboardChannel != null) {
                            updateModel.leaderboardChannel = arguments.leaderboardChannel!!.id.value.toLong()
                        }

                        if (arguments.eventActive != null) {
                            updateModel.eventActive = arguments.eventActive
                        }

                        val updatedCarryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].updateCarryType(carryType.id, updateModel)
                                ?: throw CommandExecutionWarning("Couldn't update carry type.")

                        val embed = ApplicationService.getCarryTypeEmbed(updatedCarryType)
                        embed.title = "Updated Carry Type"
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTypeResetArguments) {
                name = "reset".toKey()
                description = "Reset properties of a carry type".toKey()

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("That carry type doesn't exists!")

                        if (!arguments.logChannel && !arguments.leaderboardChannel) {
                            throw CommandExecutionWarning("Please provide something you want to reset.")
                        }

                        val updateModel = carryType.getUpdateModel()

                        if (arguments.logChannel) {
                            updateModel.logChannel = -1L
                        }

                        if (arguments.leaderboardChannel) {
                            updateModel.leaderboardChannel = -1L
                        }

                        val updatedCarryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].updateCarryType(carryType.id, updateModel)
                                ?: throw CommandExecutionWarning("Couldn't update carry type.")

                        val embed = ApplicationService.getCarryTypeEmbed(updatedCarryType)
                        embed.title = "Updated Carry Type with reset values"
                        embeds = mutableListOf(embed)
                    }
                }
            }
        }
    }

    inner class CarryTypeCreateArguments : Arguments() {
        val identifier by string {
            name = "identifier".toKey()
            description = "The identifier of the carry type.".toKey()
            maxLength = 30
        }

        val displayName by string {
            name = "display-name".toKey()
            description = "The display name of the carry type".toKey()
            maxLength = 30
        }

        val logChannel by optionalChannel {
            name = "log-channel".toKey()
            description = "Set the channel that will be used for logging".toKey()
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val leaderboardChannel by optionalChannel {
            name = "leaderboard-channel".toKey()
            description = "Set the channel that will be used to show a static leaderboard".toKey()
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val eventActive by optionalBoolean {
            name = "event-active".toKey()
            description = "Set if there if an active event for score".toKey()
        }
    }

    inner class CarryTypeArguments : Arguments() {
        val carryType by string {
            name = "carry-type".toKey()
            description = "The identifier of the carry type".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }
    }

    inner class CarryTypeEditArguments : Arguments() {
        val carryType by string {
            name = "carry-type".toKey()
            description = "The identifier of the carry type".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val displayName by optionalString {
            name = "display-name".toKey()
            description = "Set the display name".toKey()
            maxLength = 30
        }

        val logChannel by optionalChannel {
            name = "log-channel".toKey()
            description = "Set the channel that will be used for logging".toKey()
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val leaderboardChannel by optionalChannel {
            name = "leaderboard-channel".toKey()
            description = "Set the channel that will be used to show a static leaderboard".toKey()
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val eventActive by optionalBoolean {
            name = "event-active".toKey()
            description = "Set if there if an active event for score".toKey()
        }
    }

    inner class CarryTypeResetArguments : Arguments() {
        val carryType by string {
            name = "carry-type".toKey()
            description = "The identifier of the carry type".toKey()
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val logChannel by boolean {
            name = "log-channel".toKey()
            description = "Set if the log channel should be reset".toKey()
        }

        val leaderboardChannel by boolean {
            name = "leaderboard-channel".toKey()
            description = "Set if the leaderboard channel should be reset".toKey()
        }
    }
}